llumnix/queue/zmq_client.py

```diff
@@ -21,105 +22,160 @@
 import cloudpickle
 
 from llumnix.logging.logger import init_logger
+from llumnix.queue.queue_client_base import QueueClientBase
 from llumnix.server_info import ServerInfo
 
-from llumnix.queue.zmq_utils import (RPC_SUCCESS_STR, RPC_REQUEST_TYPE, RPCClientClosedError,
-                                     RPCUtilityRequest, RPCPutNoWaitQueueRequest, RPCPutNoWaitBatchQueueRequest,
-                                     get_open_zmq_ipc_path)
-from llumnix.constants import RPC_GET_DATA_TIMEOUT_MS, RPC_SOCKET_LIMIT_CUTOFF, RPC_ZMQ_HWM
+from llumnix.queue.zmq_utils import (
+    RPC_SUCCESS_STR,
+    RPC_REQUEST_TYPE,
+    RPCUtilityRequest,
+    RPCPutNoWaitQueueRequest,
+    RPCPutNoWaitBatchQueueRequest,
+    get_open_zmq_ipc_path,
+    get_zmq_pool_name,
+)
+from llumnix.constants import (
+    RPC_GET_DATA_TIMEOUT_MS,
+    SOCKET_POOL_MAXSIZE
+)
 from llumnix.metrics.timestamps import set_timestamp
 
 logger = init_logger(__name__)
 
 
-class ZmqClient:
+class AsyncZmqSocketPool:
+    def __init__(self, context: zmq.asyncio.Context, ip, port, pool_size=10):
+        self.context = context
+        self.ip = ip
+        self.port = port
+
+        # max size of the pool
+        # sockets in the pool will be long live
+        # sockets out the poll will be closed after used
+        self.pool_size = pool_size
+
+        self.pool = Queue(maxsize=pool_size)
+
+    async def _create_socket(self) -> zmq.asyncio.Socket:
+        socket = self.context.socket(zmq.DEALER)
+        dst_address = get_open_zmq_ipc_path(self.ip, self.port)
+        socket.connect(dst_address)
+        return socket
+
+    async def get_socket(self) -> zmq.asyncio.Socket:
+        socket = None
+        try:
+            socket = self.pool.get_nowait()
+        except QueueEmpty:
+            socket = await self._create_socket()
+        return socket
+
+    def full(self):
+        return self.pool.full()
+
+    def put(self, socket: zmq.asyncio.Socket):
+        self.pool.put_nowait(socket)
+
+    async def close_all_connections(self):
+        while not self.pool.empty():
+            socket = await self.pool.get()
+            socket.close(linger=0)
+
+
+class ZmqSocketPoolFactory:
+
+    def __init__(self, context: zmq.asyncio.Context, pool_size=10):
+        self.context: zmq.asyncio.Context = context
+        self.pool_size: int = pool_size
+        self.pools: Dict[str, AsyncZmqSocketPool] = {}
+        self.lock: asyncio.Lock = Lock()
+
+    async def get_pool(self, ip, port) -> AsyncZmqSocketPool:
```

Comment:
why this is a async function? why there is a asyncio lock here? is there any corotine-level concurrency here?